### DISCUSSION TITLE
Why we are not adding  Model::except(['column']) to get all columns except mentioned column(s) in Model::except()

### DISCUSSION THREAD

**Discussion by mdariftiens** [laravel/framework#51402] [Ideas]
We are getting all column by model::first() or model::get(); and specific column by passing param as column name. 
Why we are not adding  Model::except(['column']) to get all columns except mentioned column(s) in Model::except() .

**Comment by ghost** [comment]
You can simply array_diff when calling the get... Why add a method for that?

Update

select, get and distinct have complicated implementations in Eloquent. I would stick to them as they are...

Update 2

Retrieving partial models form db can mess up things later along the way as those properties will not be filled in the hydrated model.

Update 3

**In the Query\Builder there is no model so a macro is not doable.** I think if this could be resolved then @mdariftiens your wish could become a reality.

Workaround for BlogModel::except(['description'])->get();

```
class BlogModel extends Model
{
....

    /**
     * @source https://github.com/macropay-solutions/laravel-crud-wizard-free
     * MacropaySolutions\LaravelCrudWizard\Models\BaseModel::getColumns
    */
    public function getColumns(bool $includingPrimary = true): array
    {
        $columns = $includingPrimary ?
            /** $this->primaryKey can be null or empty string so, it must not be included */
            \array_merge(\array_filter([$this->primaryKey]), $this->fillable) :
            /** $this->primaryKey may be in the fillable so, it must be included */
            \array_diff($this->fillable, \array_diff([$this->primaryKey], $this->fillable));

        return \array_unique($columns);
    }

    public static function except(array $columns): Builder
    {
        $model = new static();

        return $model::query()->select(\array_values(\array_diff($model->getColumns(), $columns)));
    }
}
```

**Reply by mdariftiens** [reply]
Example 1 - `blog` table has id, title, status, description etc. `description` column may have long long data.  If we write BlogModel::get() will pick large data set with description column. If we want to pick all column except `description` we can write `BlogModel::except('description')` instead of `BlogModel::select(['id', 'title', 'status']);` . 

Example 2 - suppose table `account_detail` table has more then 50 column. now if I want to pick 40 columns I want to write 
like `AccountDetail::except([10 columns here.])->get() `instead of `AccountDetail::select([40 columns here.])->get() `

please suggest me if there any other way I can use.

**Reply by macropay-solutions** [reply]
https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/36312b5dbd9c06aba06714a182397303eced1ffa/src/Models/BaseModel.php#L39

BlogModel::query()->get(\array_values(\array_diff((new BlogModel())->getColumns(), ['description'])));

**Reply by mdariftiens** [reply]
I think `AccountDetail::except(['description'])->get()` is easier than `BlogModel::query()->get(\array_diff((new BlogModel())->getColumns(), ['description']));` . 

@nunomaduro want you opinion also.

**Reply by ghost** [reply]
@mdariftiens eloquent uses * and will not compile all your columns from fillable. 

A possible solution is to create that except function as a macro that will be a wrapper over what @macropay-solutions posted.

Update: macro is not doable because in Query\Builder the model is not accessible.
